lib/Conversion/TritonGPUToLLVM/DotOpToLLVM/MMAv3.cpp

```diff
@@ -0,0 +1,209 @@
+#ifdef USE_ROCM
+
+#include "../DotOpToLLVM.h"
+#include "../Utility.h"
+
+#include "mlir/Dialect/LLVMIR/ROCDLDialect.h"
+
+using namespace mlir;
+using namespace mlir::triton;
+
+namespace {
+
+using ::mlir::triton::gpu::DotOperandEncodingAttr;
+using ::mlir::triton::gpu::MmaEncodingAttr;
+using ::mlir::triton::gpu::SharedEncodingAttr;
+
+enum class MatrixCoreType : uint8_t {
+  // D = AB + C
+  FP32_FP16_FP16_FP32 = 0, // default
+  FP32_BF16_BF16_FP32,
+  FP32_FP32_FP32_FP32,
+  FP64_FP64_FP64_FP64,
+  INT32_INT8_INT8_INT32,
+  NOT_APPLICABLE,
+};
+
+using ValueTable = std::map<std::pair<unsigned, unsigned>, Value>;
+
+struct DotOpMFMAConversionHelper {
+  MmaEncodingAttr mmaLayout;
+
+  ConversionPatternRewriter &rewriter;
+  TritonGPUToLLVMTypeConverter *typeConverter;
+  Location loc;
+  MLIRContext *ctx{};
+
+  explicit DotOpMFMAConversionHelper(
+      MmaEncodingAttr mmaLayout, ConversionPatternRewriter &rewriter,
+      TritonGPUToLLVMTypeConverter *typeConverter, Location loc)
+      : mmaLayout(mmaLayout), rewriter(rewriter), typeConverter(typeConverter),
+        loc(loc), ctx(mmaLayout.getContext()) {}
+
+  Value getThreadId() const {
+    auto llvmIndexTy = typeConverter->getIndexType();
+    auto tid = rewriter.create<::mlir::gpu::ThreadIdOp>(
+        loc, rewriter.getIndexType(), ::mlir::gpu::Dimension::x);
+    return rewriter.create<arith::TruncIOp>(loc, i32_ty, tid);
+  }
+
+  Value generateMFMAOp(MatrixCoreType mfmaTy, Value valA, Value valB,
+                       Value valC) const {
+    auto resType = valC.getType();
+    Value zeroFlag = i32_val(0);
+    switch (mfmaTy) {
+    case MatrixCoreType::FP32_FP16_FP16_FP32:
+      return rewriter.create<ROCDL::mfma_f32_32x32x8f16>(
```

Comment:
What is the plan for other shape types e.g. mfma_f32_16x16x16f16?

Reply:
Unfortunately, current compilation pipeline does not provide an easy way to switch between different tile sizes "on the fly".
First we need to measure performance of different operations for target matrix sizes different instruction shapes provides some advantages,